packages/router-core/src/new-process-route-tree.ts

```diff
@@ -0,0 +1,614 @@
+		skipped: number
+	}
+
+	// use a stack to explore all possible paths (optional params cause branching)
+	// we use a depth-first search, return the first result found
+	const stack: Array<Frame> = [
+		{ node: segmentTree, index: 0, depth: 0, skipped: 0 }
+	]
+
+	let wildcardMatch: Frame | null = null
+
+	while (stack.length) {
+		// eslint-disable-next-line prefer-const
+		let { node, index, skipped, depth } = stack.shift()!
+
+		main: while (node && index <= parts.length) {
+			if (index === parts.length) {
+				if (!node.routeId) break
+				return { node, skipped }
+			}
+
+			const part = parts[index]!
+
+			// 1. Try static match
+			if (node.static) {
+				const match = node.static.get(part)
+				if (match) {
+					node = match
+					depth++
+					index++
+					continue
+				}
+			}
+
+			// 2. Try case insensitive static match
+			if (node.staticInsensitive) {
+				const match = node.staticInsensitive.get(part.toLowerCase())
+				if (match) {
+					node = match
+					depth++
+					index++
+					continue
+				}
+			}
+
+			// 3. Try dynamic match
+			if (node.dynamic) {
+				for (const segment of node.dynamic) {
+					const { prefix, suffix } = segment
+					if (prefix || suffix) {
+						const casePart = segment.caseSensitive ? part : part.toLowerCase()
+						if (prefix && !casePart.startsWith(prefix)) continue
+						if (suffix && !casePart.endsWith(suffix)) continue
+					}
+					node = segment
+					depth++
+					index++
+					continue main
+				}
+			}
+
+			// 4. Try optional match
+			if (node.optional) {
+				const nextDepth = depth + 1
+				const nextSkipped = skipped | (1 << nextDepth)
+				for (const segment of node.optional) {
+					// when skipping, node and depth advance by 1, but index doesn't
+					stack.push({ node: segment, index, skipped: nextSkipped, depth: nextDepth }) // enqueue skipping the optional
+				}
+				for (const segment of node.optional) {
+					const { prefix, suffix } = segment
+					if (prefix || suffix) {
+						const casePart = segment.caseSensitive ? part : part.toLowerCase()
+						if (prefix && !casePart.startsWith(prefix)) continue
+						if (suffix && !casePart.endsWith(suffix)) continue
+					}
+					node = segment
+					index++
+					depth++
+					continue main
+				}
+			}
+
+			// 5. Try wildcard match
+			if (node.wildcard) {
+				for (const segment of node.wildcard) {
+					const { prefix, suffix } = segment
+					if (prefix) {
+						const casePart = segment.caseSensitive ? part : part.toLowerCase()
+						if (!casePart.startsWith(prefix)) continue
+					}
+					if (suffix) {
+						const part = parts[parts.length - 1]!
+						const casePart = segment.caseSensitive ? part : part.toLowerCase()
+						if (!casePart.endsWith(suffix)) continue
+					}
+					// a wildcard match terminates the loop, but we need to continue searching in case there's a longer match
+					if (!wildcardMatch || (wildcardMatch.index < index)) {
+						wildcardMatch = { node: segment, index, skipped, depth }
+					}
```

Comment:
score should be on `DynamicNode`, used in sort above, and in here if both indices are equal.